fendermint/vm/topdown/proof-service/src/cache.rs

```diff
@@ -0,0 +1,372 @@
+
+        // Load all entries from disk into memory
+        let entries_vec = persistence.load_all_entries()?;
+        let entries: BTreeMap<u64, CacheEntry> = entries_vec
+            .into_iter()
+            .map(|e| (e.instance_id, e))
+            .collect();
+
+        tracing::info!(
+            last_committed,
+            entry_count = entries.len(),
+            "Loaded cache from disk"
+        );
+
+        Ok(Self {
+            entries: Arc::new(RwLock::new(entries)),
+            last_committed_instance: Arc::new(AtomicU64::new(last_committed)),
+            config,
+            persistence: Some(Arc::new(persistence)),
+        })
+    }
+
+    /// Get the next uncommitted proof (in sequential order)
+    /// Returns the proof for (last_committed + 1)
+    pub fn get_next_uncommitted(&self) -> Option<CacheEntry> {
+        let last_committed = self.last_committed_instance.load(Ordering::Acquire);
+        let next_instance = last_committed + 1;
+
+        self.entries.read().get(&next_instance).cloned()
+    }
+
+    /// Get proof for a specific instance ID
+    pub fn get(&self, instance_id: u64) -> Option<CacheEntry> {
+        self.entries.read().get(&instance_id).cloned()
+    }
+
+    /// Check if an instance is already cached
+    pub fn contains(&self, instance_id: u64) -> bool {
+        let result = self.entries.read().contains_key(&instance_id);
+
+        // Record cache hit/miss
+        use crate::observe::CACHE_HIT_TOTAL;
+        CACHE_HIT_TOTAL
+            .with_label_values(&[if result { "hit" } else { "miss" }])
+            .inc();
+
+        result
+    }
+
+    /// Insert a proof into the cache
+    pub fn insert(&self, entry: CacheEntry) -> anyhow::Result<()> {
+        let instance_id = entry.instance_id;
+
+        // Check if we're within the lookahead window
+        let last_committed = self.last_committed_instance.load(Ordering::Acquire);
+        let max_allowed = last_committed + self.config.lookahead_instances;
+
+        if instance_id > max_allowed {
+            anyhow::bail!(
+                "Instance {} exceeds lookahead window (last_committed={}, max={})",
+                instance_id,
+                last_committed,
+                max_allowed
+            );
+        }
+
+        // Insert to memory
+        self.entries.write().insert(instance_id, entry.clone());
+
+        // Persist to disk if enabled
+        if let Some(persistence) = &self.persistence {
+            persistence.save_entry(&entry)?;
+        }
+
+        // Emit metrics
+        let cache_size = self.entries.read().len();
+        let highest = self.highest_cached_instance();
+
+        if let Some(highest_cached) = highest {
+            emit(ProofCached {
+                instance: instance_id,
+                cache_size,
+                highest_cached,
+            });
+        }
+
+        // Update cache size metric
+        use crate::observe::CACHE_SIZE;
+        CACHE_SIZE.set(cache_size as i64);
```

Comment:
Should do the same in `cleanup_old_instances`

Reply:
We should.